### PR TITLE
Installer updates

### PR DESCRIPTION
**Pull request by dgoodwin**
Various changes on getting hive to work with a live install config, and cleanup more reliably.

### CODE REVIEW

**Comment by dgoodwin** [comment]
@csrwng how is this looking to you?

Many settings in install-config still won't take effect but we're getting closer.

File: contrib/pkg/installmanager/installmanager.go [hunk 1]
```diff
@@ -302,6 +313,26 @@ func (m *InstallManager) uploadAdminKubeconfig() error {
...
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, in)
+	if err != nil {
+		return err
+	}
+	return out.Close()
```
Comment (csrwng) [inline contrib/pkg/installmanager/installmanager.go]: You're already deferring out.Close() above... why call it again?

Comment (csrwng) [reply contrib/pkg/installmanager/installmanager.go]: You may want to call out.Sync() instead to make sure it's flushed to disk

File: contrib/pkg/installmanager/installmanager_test.go [hunk 2]
```diff
@@ -141,6 +148,13 @@ func writeFakeBinary(fileName string, contents string) error {
 	return err
 }
 
+func writeFakeInstallConfig(fileName string) error {
+	// nothing needs to read this so for now just an empty file
+	data := []byte("fakefile")
+	err := ioutil.WriteFile(fileName, data, 0755)
```
Comment (csrwng) [inline contrib/pkg/installmanager/installmanager_test.go]: nit: `return ioutil.WriteFile(fileName, data, 0755)`

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 3]
```diff
@@ -136,6 +136,10 @@ func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (recon
 	cdLog.Info("reconciling cluster deployment")
 	cd = cd.DeepCopy()
 
+	if cd.Status.ClusterUUID == "" {
```
Comment (csrwng) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go]: I'm wondering if ClusterUUID really belongs in Status. Is there ever a case when the end-user wants to specify a ClusterUUID? (to adopt an existing cluster, for example)

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 4]
```diff
@@ -343,7 +319,49 @@ func (r *ReconcileClusterDeployment) updateClusterDeploymentStatus(cd *hivev1.Cl
...
+
+	if cd.Status.ClusterUUID != "" {
+		return fmt.Errorf("cluster UUID already set")
+	}
+
+	cd.Status.ClusterUUID = uuid.New()
+	cdLog.WithField("clusterUUID", cd.Status.ClusterUUID).Info("generated new cluster UUID")
+	err := r.Update(context.TODO(), cd)
+	if err != nil {
+		cdLog.Errorf("error updating cluster deployment: %v", err)
```
Comment (csrwng) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go]: use `WithError`

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 5]
```diff
@@ -343,7 +319,49 @@ func (r *ReconcileClusterDeployment) updateClusterDeploymentStatus(cd *hivev1.Cl
...
+	// Delete the install job in case it's still running:
+	installJob := &batchv1.Job{}
+	err := r.Get(context.Background(),
+		types.NamespacedName{
+			Name:      install.GetInstallJobName(cd),
+			Namespace: cd.Namespace,
+		},
+		installJob)
+	if err != nil && errors.IsNotFound(err) {
+		cdLog.Debugf("install job no longer exists, nothing to cleanup")
```
Comment (csrwng) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go]: s/Debugf/Debug/

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go [hunk 6]
```diff
@@ -111,25 +112,27 @@ func TestClusterDeploymentReconcile(t *testing.T) {
...
 				testSecret(adminPasswordSecret, adminCredsSecretPasswordKey, "password"),
 				testSecret(pullSecretSecret, pullSecretKey, "{}"),
 				testSecret(sshKeySecret, adminSSHKeySecretKey, "fakesshkey"),
 			},
-			validate: func(c client.Client) {
+			validate: func(c client.Client, t *testing.T) {
 				cd := getCD(c)
 				if cd == nil || !apiequality.Semantic.DeepEqual(cd, testClusterDeployment()) {
+					t.Logf("%v", cd)
+					t.Logf("%v", testClusterDeployment())
```
Comment (csrwng) [inline pkg/controller/clusterdeployment/clusterdeployment_controller_test.go]: Are these log stmts intentional?

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go [hunk 7]
```diff
@@ -197,7 +197,7 @@ func TestClusterDeploymentReconcile(t *testing.T) {
 				testSecret(pullSecretSecret, pullSecretKey, "{}"),
 				testSecret(sshKeySecret, adminSSHKeySecretKey, "fakesshkey"),
 			},
-			validate: func(c client.Client) {
+			validate: func(c client.Client, t *testing.T) {
```
Comment (csrwng) [inline pkg/controller/clusterdeployment/clusterdeployment_controller_test.go]: In this delete case, validate that the job was deleted as well (add one in existing objects)

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 3 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go]: If it were moved to spec, is it ok for us to automatically fill that out if unset (which it normally would be?) I guess that would probably be fine.

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go (hunk 7 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller_test.go]: Good catch. I couldn't add here as this just triggers a re-sync but it seems to fit above in the more straightforward delete test.

**Pull request merged**